### PR TITLE
[Task 1837495] use mysql-lib in intellij plugin

### PR DESCRIPTION
**Pull request by andxu**
[AB#1837495](https://dev.azure.com/mseng/a4d27ce2-a42d-4b71-8eef-78cee9a9728e/_workitems/edit/1837495)

### CODE REVIEW

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/mysql/MySQLNode.java [hunk 1]
```diff
@@ -23,37 +23,56 @@
...
-        this.serverState = server.userVisibleState();
+        this.serverState = server.entity().getState();
         loadActions();
+        AzureEventBus.after("mysql|server.restart", this::onMySqlServerStatusChanged);
+        AzureEventBus.after("mysql|server.stop", this::onMySqlServerStatusChanged);
+        AzureEventBus.after("mysql|server.restart", this::onMySqlServerStatusChanged);
+        AzureEventBus.before("mysql|server.start", this::onMySqlServerStatusChanging);
+        AzureEventBus.before("mysql|server.stop", this::onMySqlServerStatusChanging);
+        AzureEventBus.before("mysql|server.restart", this::onMySqlServerStatusChanging);
+        AzureEventBus.before("mysql|server.delete", this::onMySqlServerStatusChanging);
```
Comment (shenqianjin) [inline Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/mysql/MySQLNode.java:57]: line 51/53 are duplicated

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/mysql/MySQLNode.java [hunk 2]
```diff
@@ -76,8 +95,9 @@ private BasicActionBuilder initActionBuilder(Runnable runnable) {
 
     @Override
     public List<NodeAction> getNodeActions() {
-        boolean updating = SERVER_UPDATING.equals(serverState);
-        boolean running = ServerState.READY.equals(serverState);
+        boolean running = StringUtils.equalsIgnoreCase("READY", serverState);
+        boolean updating = StringUtils.equalsIgnoreCase(SERVER_UPDATING, serverState);
```
Comment (shenqianjin) [inline Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/mysql/MySQLNode.java:99]: make status such as "READY" as a variable ?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/mysql/SpringDatasourceLineMarkerProvider.java [hunk 3]
```diff
@@ -87,9 +88,9 @@ public void navigate(MouseEvent mouseEvent, PsiElement psiElement) {
                 return;
             }
             final ResourceId serverId = database.getServerId();
-            final Server server = MySQLMvpModel.findServer(serverId.subscriptionId(), serverId.resourceGroupName(), serverId.name());
+            final MySqlServer server = Azure.az(AzureMySql.class).subscription(serverId.subscriptionId()).get(serverId.id());
             if (Objects.nonNull(server)) {
```
[the remark this answers is not in the document]
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/mysql/SpringDatasourceLineMarkerProvider.java:92]: the resource id was brought by Miller

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/mysql/MySQLNode.java (hunk 2 above)
Comment (andxu) [reply Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/mysql/MySQLNode.java:99]: should be laid in later refactor , not track2 refactor

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/lib/mysql/AzureMySQLConfig.java [hunk 4]
```diff
@@ -47,22 +48,20 @@
     public static AzureMySQLConfig getDefaultAzureMySQLConfig() {
         final String defaultNameSuffix = DateFormatUtils.format(new Date(), "yyyyMMddHHmmss");
         final AzureMySQLConfig config = new AzureMySQLConfig();
-        List<Subscription> selectedSubscriptions = az(AzureAccount.class).account().getSelectedSubscriptions();
-        Preconditions.checkArgument(CollectionUtils.isNotEmpty(selectedSubscriptions), "There is no subscription in your account.");
-        Subscription subscription = selectedSubscriptions.get(0);
+        final List<Subscription> selectedSubscriptions = az(AzureAccount.class).account().getSelectedSubscriptions();
+        Preconditions.checkArgument(CollectionUtils.isNotEmpty(selectedSubscriptions), "There are no subscriptions in your account.");
```
[the remark this answers is not in the document]
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/lib/mysql/AzureMySQLConfig.java:52]: image

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/lib/mysql/AzureMySQLConfig.java [hunk 5]
```diff
@@ -47,22 +48,20 @@
...
-        }
-        List<Region> supportedRegions = MySQLMvpModel.listSupportedRegions(subscription);
+        final AzureMySql azureMySql = az(AzureMySql.class).subscription(subscription.getId());
+        final List<Region> supportedRegions = azureMySql.listSupportedRegions();
         if (CollectionUtils.isNotEmpty(supportedRegions)) {
             config.setRegion(supportedRegions.get(0));
         }
         config.setServerName("mysql-" + defaultNameSuffix);
         config.setAdminUsername(StringUtils.EMPTY);
+        config.setVersion("5.7"); // defeault to 5.7
```
[the remark this answers is not in the document]
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/lib/mysql/AzureMySQLConfig.java:64]: default version in lib is null, actually 5.7

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/mysql/MySQLNode.java (hunk 1 above)
Comment (andxu) [reply Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/mysql/MySQLNode.java:57]: onMySqlServerStatusChanged onMySqlServerStatusChanging

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/mysql/MySQLModule.java [hunk 6]
```diff
@@ -26,12 +27,18 @@
     public MySQLModule(final Node parent) {
         super(MYSQL_DATABASE_MODULE_ID, MODULE_NAME, parent);
         createListener();
+
+        AzureEventBus.after("mysql|server.create", this::onMySqlServerCreatedOrRemoved);
+        AzureEventBus.after("mysql|server.delete", this::onMySqlServerCreatedOrRemoved);
+    }
```
Comment (shenqianjin) [inline Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/mysql/MySQLModule.java:33]: how about replace "this::onMySqlServerCreatedOrRemoved" with "e -> this.refreshItems()" ?

**Review by shenqianjin — APPROVED** [review]

File: PluginsAndFeatures/azure-toolkit-for-intellij/resources/com/microsoft/azure/toolkit/operation/titles_mysql_server_intellij.properties [hunk 7]
```diff
@@ -0,0 +1,5 @@
+mysql|server.create=create MySQL server({})
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/resources/com/microsoft/azure/toolkit/operation/titles_mysql_server_intellij.properties:1]: change to {0}

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/mysql/component/DatabaseComboBox.java [hunk 8]
```diff
@@ -5,33 +5,31 @@
...
     @Getter
-    private Server server;
+    private MySqlServer server;
 
     public DatabaseComboBox() {
         super(false);
     }
 
-    public void setServer(Server server) {
+    public void setServer(MySqlServer server) {
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/mysql/component/DatabaseComboBox.java:27]: add @Nullable

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/mysql/component/DatabaseComboBox.java [hunk 9]
```diff
@@ -40,21 +38,23 @@ public void setServer(Server server) {
...
         }
         return super.getItemText(item);
     }
 
     @Override
-    protected List<? extends DatabaseInner> loadItems() {
-        final AzureManager manager = AuthMethodManager.getInstance().getAzureManager();
-        if (Objects.isNull(server) || Objects.isNull(manager)) {
+    @AzureOperation(
+        name = "mysql|database.list.server|subscription",
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/mysql/component/DatabaseComboBox.java:49]: remove `|subscription`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/DatabaseComboBox.java [hunk 10]
```diff
@@ -52,23 +51,25 @@ public void setServer(Server server) {
...
-            return ((DatabaseInner) item).name();
+        if (item instanceof MySqlDatabaseEntity) {
+            return ((MySqlDatabaseEntity) item).getName();
         }
         return super.getItemText(item);
     }
 
     @Override
     @AzureOperation(
         name = "mysql|database.list.server|subscription",
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/DatabaseComboBox.java:62]: remove |subscription

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/action/CreateMySQLAction.java [hunk 11]
```diff
@@ -74,20 +87,58 @@ private void createAzureMySQL(final AzureMySQLConfig config, final Project proje
...
+                .withVersion(config.getVersion())
+                .withAdministratorLogin(config.getAdminUsername())
+                .withAdministratorLoginPassword(String.valueOf(config.getPassword())).commit();
+            // update access from azure services
+            if (config.isAllowAccessFromAzureServices()) {
+                server.firewallRules().enableAzureAccessRule();
+            }
+            // update access from local machine
+            if (config.isAllowAccessFromLocalMachine()) {
+                server.firewallRules().enableLocalMachineAccessRule(server.getPublicIpForLocalMachine());
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/mysql/action/CreateMySQLAction.java:134]: remove parameter or change to use 'addFirewallRule'

File: Utils/azure-explorer-common/pom.xml [hunk 12]
```diff
@@ -11,6 +11,8 @@
     <properties>
         <artifact.name>${project.artifactId}-${project.version}.jar</artifact.name>
         <checkstyle.skip>true</checkstyle.skip>
+        <maven.compiler.source>11</maven.compiler.source>
```
Comment (Flanker32) [inline Utils/azure-explorer-common/pom.xml:14]: Move to utils later

**Review by Flanker32 — APPROVED** [review]

**Pull request merged**